File: src/test/java/no/novari/configuration/security/TokenParsingUtilsTest.java
package no.novari.configuration.security;

import no.novari.configuration.model.User;
import no.novari.configuration.security.TokenParsingUtils;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;

import java.util.Map;
import java.util.Optional;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class TokenParsingUtilsTest {

    private TokenParsingUtils tokenParsingUtils;

    @BeforeEach
    void setUp() {
        tokenParsingUtils = new TokenParsingUtils();
    }

    @Test
    void shouldReturnEmptyWhenAuthenticationIsNotJwt() {
        Authentication authentication = mock(Authentication.class);

        Optional<User> result = tokenParsingUtils.tryGetUser(authentication);

        assertThat(result).isEmpty();
    }

    @Test
    void shouldReturnEmptyWhenAttributesAreMissing() {
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of());

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);

        assertThat(result).isEmpty();
    }

    @Test
    void shouldReturnUserWhenDisplayNameEmailAndOidArePresent() {
        UUID oid = UUID.randomUUID();
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of(
                "displayname", "John Doe",
                "email", "john@example.com",
                "objectidentifier", oid.toString()
        ));

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);

        assertThat(result).isPresent();
        User user = result.get();
        assertThat(user.getName()).isEqualTo("John Doe");
        assertThat(user.getEmail()).isEqualTo("john@example.com");
        assertThat(user.getObjectIdentifier()).isEqualTo(oid);
    }

    @Test
    void shouldReturnEmptyWhenEmailIsMissing() {
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of(
                "displayname", "Jane Smith",
                "principalName", "jane.smith@org.no"
        ));

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);

        assertThat(result).isEmpty();
    }

    @Test
    void shouldReturnEmptyWhenOidIsInvalid() {
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of(
                "displayname", "Broken UUID",
                "objectidentifier", "not-a-uuid"
        ));

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);

        assertThat(result).isEmpty();
    }

    @Test
    void shouldParseUserWhenOnlyEmailAndOidArePresent() {
        UUID oid = UUID.randomUUID();
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of(
                "email", "test@domain.no",
                "objectidentifier", oid.toString()
        ));

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);

        assertThat(result).isPresent();
        User user = result.get();
        assertThat(user.getName()).isNull();
        assertThat(user.getEmail()).isEqualTo("test@domain.no");
        assertThat(user.getObjectIdentifier()).isEqualTo(oid);
    }

    @Test
    void shouldIgnoreBlankOrWhitespaceAttributes() {
        UUID oid = UUID.randomUUID();
        JwtAuthenticationToken jwt = mock(JwtAuthenticationToken.class);
        when(jwt.getTokenAttributes()).thenReturn(Map.of(
                "displayname", "  ",
                "email", " ",
                "objectidentifier", oid.toString()
        ));

        Optional<User> result = tokenParsingUtils.tryGetUser(jwt);
        assertThat(result).isEmpty();
    }

    @Test
    void shouldConvertToTitleCaseCorrectly() {
        String input = "john DOE";
        String result = invokeToTitleCase(input);
        assertThat(result).isEqualTo("John Doe");
    }

    @Test
    void shouldPreserveNonLetterCharactersInTitleCase() {
        String input = "anne-marie o’neill";
        String result = invokeToTitleCase(input);
        assertThat(result).isEqualTo("Anne-Marie O’Neill");
    }

    private String invokeToTitleCase(String input) {
        try {
            var method = TokenParsingUtils.class.getDeclaredMethod("toTitleCase", String.class);
            method.setAccessible(true);
            return (String) method.invoke(tokenParsingUtils, input);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/test/kotlin/no/novari/flyt/configuration/security/TokenParsingUtilsTest.kt
package no.novari.flyt.configuration.security

import no.novari.flyt.configuration.model.User
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.springframework.security.core.Authentication
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import java.util.UUID

class TokenParsingUtilsTest {
    private lateinit var tokenParsingUtils: TokenParsingUtils

    @BeforeEach
    fun setUp() {
        tokenParsingUtils = TokenParsingUtils()
    }

    @Test
    fun shouldReturnEmptyWhenAuthenticationIsNotJwt() {
        val authentication = mock<Authentication>()

        val result: User? = tokenParsingUtils.tryGetUser(authentication)

        assertThat(result).isNull()
    }

    @Test
    fun shouldReturnEmptyWhenAttributesAreMissing() {
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(mapOf())

        val result: User? = tokenParsingUtils.tryGetUser(jwt)

        assertThat(result).isNull()
    }

    @Test
    fun shouldReturnUserWhenDisplayNameEmailAndOidArePresent() {
        val oid = UUID.randomUUID()
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "displayname" to "John Doe",
                "email" to "john@example.com",
                "objectidentifier" to oid.toString(),
            ),
        )

        val result = tokenParsingUtils.tryGetUser(jwt)

        val user = requireNotNull(result)
        assertThat(user.name).isEqualTo("John Doe")
        assertThat(user.email).isEqualTo("john@example.com")
        assertThat(user.objectIdentifier).isEqualTo(oid)
    }

    @Test
    fun shouldReturnEmptyWhenEmailIsMissing() {
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "displayname" to "Jane Smith",
                "principalName" to "jane.smith@org.no",
            ),
        )

        val result = tokenParsingUtils.tryGetUser(jwt)

        assertThat(result).isNull()
    }

    @Test
    fun shouldReturnEmptyWhenOidIsInvalid() {
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "displayname" to "Broken UUID",
                "objectidentifier" to "not-a-uuid",
            ),
        )

        val result = tokenParsingUtils.tryGetUser(jwt)

        assertThat(result).isNull()
    }

    @Test
    fun shouldParseUserWhenOnlyEmailAndOidArePresent() {
        val oid = UUID.randomUUID()
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "email" to "test@domain.no",
                "objectidentifier" to oid.toString(),
            ),
        )

        val result = tokenParsingUtils.tryGetUser(jwt)

        val user = requireNotNull(result)
        assertThat(user.name).isNull()
        assertThat(user.email).isEqualTo("test@domain.no")
        assertThat(user.objectIdentifier).isEqualTo(oid)
    }

    @Test
    fun shouldIgnoreBlankOrWhitespaceAttributes() {
        val oid = UUID.randomUUID()
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "displayname" to "  ",
                "email" to " ",
                "objectidentifier" to oid.toString(),
            ),
        )

        val result = tokenParsingUtils.tryGetUser(jwt)
        assertThat(result).isNull()
    }

    @Test
    fun shouldReturnAuditorNameWhenOidIsMissing() {
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "displayname" to "Jane Smith",
                "email" to "jane.smith@org.no",
            ),
        )

        val result = tokenParsingUtils.tryGetAuditorName(jwt)

        assertThat(result).isEqualTo("Jane Smith")
    }

    @Test
    fun shouldReturnEmailAsAuditorNameWhenOnlyPrincipalNameIsPresent() {
        val jwt = mock<JwtAuthenticationToken>()
        whenever(jwt.tokenAttributes).thenReturn(
            mapOf(
                "principalName" to "jane.smith@org.no",
            ),
        )

        val result = tokenParsingUtils.tryGetAuditorName(jwt)

        assertThat(result).isEqualTo("jane.smith@org.no")
    }

    @Test
    fun shouldConvertToTitleCaseCorrectly() {
        val input = "john DOE"
        val result = invokeToTitleCase(input)
        assertThat(result).isEqualTo("John Doe")
    }

    @Test
    fun shouldPreserveNonLetterCharactersInTitleCase() {
        val input = "anne-marie o’neill"
        val result = invokeToTitleCase(input)
        assertThat(result).isEqualTo("Anne-Marie O’Neill")
    }

    private fun invokeToTitleCase(input: String): String =
        try {
            val method = TokenParsingUtils::class.java.getDeclaredMethod("toTitleCase", String::class.java)
            method.isAccessible = true
            method.invoke(tokenParsingUtils, input) as String
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
}
